Final Project -- Tux / bEtsy

catchingash opened the pull request:
This meets the requirements for bEtsy.

I completed this project with @drvonnjerryxlii, @lindseyreno, and @corinnepingul.

URL: https://desolate-coast-1026.herokuapp.com/

sojeri committed:
implementing changes as discussed in PR

sojeri committed:
merged in PR22

catchingash committed:
Merge pull request #20 from catchingash/tux/js/order-model-validations-specs

expanded order model validations & updated specs accordingly

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/OrdersController_expand_testing

sojeri committed:
wip fleshing out index/show pages to have customer info

lindseyreno committed:
Adds a dashboard for a seller

corinnepingul committed:
Styled the sign-up page.

lindseyreno committed:
Adds tests for a seller dashboard and revises route

lindseyreno committed:
Changes the redirect to the seller's dashboard after login

corinnepingul committed:
Added a todo comment for when we use the dashboard.

lindseyreno committed:
Adds some content to seller dashboard

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/sellers_dashboard

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/styling

lindseyreno committed:
Added back in deleted else

corinnepingul committed:
Merge pull request #23 from catchingash/tux/lr/sellers_dashboard

Tux/lr/sellers dashboard

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/styling

sojeri committed:
merged in PR23

corinnepingul committed:
added link to user's dashboard.

lindseyreno committed:
Logs in a newly signed up seller and takes them to their dashboard

lindseyreno committed:
Adds successful signup message

corinnepingul committed:
Changed page header for the dashboard.

corinnepingul committed:
Merge pull request #24 from catchingash/tux/lr/sign_up_seller

Tux/lr/sign up seller

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/styling

sojeri committed:
updated before action to redirect when logged in seller tries to access another seller's restricted content

catchingash committed:
Fixes broken specs that happened after merge.

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/OrdersController_expand_testing

corinnepingul committed:
Added comment back in.

sojeri committed:
some optimization of error messages & other minor changes

catchingash committed:
Merge pull request #25 from catchingash/tux/crp/styling

Styling

catchingash committed:
Adds TODO comment.

sojeri committed:
merged in PR25

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/OrdersController_expand_testing

sojeri committed:
updated seeds file to use smaller line sizes per class style guide

lindseyreno committed:
Fixes signup redirect test

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/OrdersController_expand_testing

sojeri committed:
missed a spot in seeds file

catchingash committed:
Refactor -- creates order#checkout method to follow single responsibility principles

catchingash committed:
Switches from an unnecessary redirect to a render.

catchingash committed:
Deletes unnecessary error message.

catchingash committed:
Refactor -- moves alert message into MESSAGES.

Commit:
Merge pull request #26 from catchingash/tux/anw/OrdersController_expand_testing

Refactors session[:order_id], OrdersController, expands testing

lindseyreno committed:
Fixes top_product test

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/bug_fixes

catchingash committed:
Merge pull request #27 from catchingash/tux/lr/bug_fixes

Rspec Bug fixes -- adds 2 products so it's top 12 (as expected) rather than top 10

sojeri committed:
adding links to dashboard

corinnepingul committed:
Changed button name from 'more information about this product' to 'details'.

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/fixing_button

sojeri committed:
merged in PR26

lindseyreno committed:
Fixes test for #index

catchingash committed:
Fixes spec that wasn't testing validity properly

catchingash committed:
Merge pull request #28 from catchingash/tux/crp/fixing_button

Fixing button

lindseyreno committed:
Merge pull request #30 from catchingash/tux/anw/OrderItemModel_bug_fix_rspec

OrderItem Model Bug -- fix rspec

corinnepingul committed:
Merge pull request #29 from catchingash/tux/lr/categories_fix

Fixes test for #index

sojeri committed:
scaffolded a little of the seller-id-products & updated products controller before_actions in process

corinnepingul committed:
Abstracted some view logic into the application helper.

catchingash committed:
Removes unnecessary before_create. The validation `order_item_is_unique?` takes care of this.

corinnepingul committed:
Switched except to only for before actions.

corinnepingul committed:
Switched except to only for before actions.

corinnepingul committed:
Changed this to a one-liner.

lindseyreno committed:
Merge pull request #32 from catchingash/tux/anw/minor_changes

Removes unnecessary before_create.

sojeri committed:
added tests for index & show views in orders controller

catchingash committed:
Merge pull request #33 from catchingash/tux/crp/switch_before_actions_to_only

Switch before actions to only

catchingash committed:
Merge pull request #31 from catchingash/tux/crp/adding_helper_methods

Abstracted some view logic into the application helper.

sojeri committed:
merged PRs 28-33

lindseyreno committed:
Merge pull request #34 from catchingash/tux/js/creating-sellers-id-orders

creating sellers id orders

sojeri committed:
added flash[:errors] specs to order_items_controller

sojeri committed:
added flash[:errors] specs to products controller

sojeri committed:
added flash[:errors] specs to reviews controller

sojeri committed:
added flash[:errors] specs to sellers controller

sojeri committed:
added flash[:errors] specs to orders controller

sojeri committed:
merged in tux/master. ready for PR.

corinnepingul committed:
Changed method name from 'order_price' to 'total_order_price' and made sure all files that call this method are also changed(including specs).

sojeri committed:
moved remove_prompt_text from OrderItem model to OrderItemHelpers

sojeri committed:
updated cart_display_text so cart link could always show up

corinnepingul committed:
Changed name of 'item_price' to 'total_item_price' in the specs and in the model and views.

Commit:
Merge pull request #36 from catchingash/tux/crp/price_method_refactor

Changed method name from 'order_price' to 'total_order_price' and mad…

lindseyreno committed:
Changes the redirect after updating a product to the seller's product page

lindseyreno committed:
Adds content to the page and adds a test for the action

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/sellers_products

lindseyreno committed:
Fixes merge conflicts

lindseyreno committed:
Merge branch 'tux/crp/total_item_price_name_change' into tux/master

sojeri committed:
added pending example spec back to helper spec files

catchingash committed:
Merge pull request #39 from catchingash/tux/lr/sellers_products

sellers products view

catchingash committed:
Merge pull request #35 from catchingash/tux/js/order-items-controller-errors-specs

adding flash[:errors] specs for all controller specs missing them

lindseyreno committed:
Refactors #total_order_price

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/refactor_total_order_price

sojeri committed:
moved spec for remove_prompt_text from order items model to order items helper

catchingash committed:
Changed `@item` --> `@order_item` for clarity.

Also deleted a subsequently failing spec that we will no longer need after future changes.
Today, we decided that we want to let buyers add items to their carts, even if someone else has the items in their cart.

lindseyreno committed:
WIP cart

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/rename_item_variable

lindseyreno committed:
Changes the cart adding the individual product price

sojeri committed:
added specs for helpers in application helpers

lindseyreno committed:
Makes the cart and checkout views match

lindseyreno committed:
Merge pull request #42 from catchingash/tux/anw/rename_item_variable

Rename item variable

catchingash committed:
Merge pull request #43 from catchingash/tux/lr/cart

Shows individual product price in cart

lindseyreno committed:
Made it one line

lindseyreno committed:
Merge pull request #41 from catchingash/tux/lr/refactor_total_order_price

Tux/lr/refactor total order price

catchingash committed:
Merge pull request #37 from catchingash/tux/js/removing-helpers-from-order-item-model

removing helper from order item model & updating cart display in navbar

catchingash committed:
Updates messages/errors so that they display appropriately.

TODO: If we have time, it'd be a good idea to write our display messages better.

Commit:
Merge pull request #44 from catchingash/tux/anw/fix_error_message_display

Updates messages/errors so that they display appropriately.

lindseyreno committed:
Fixed test. All tests green now

corinnepingul committed:
Merge pull request #45 from catchingash/tux/lr/bug_fix

Fixed test. All tests green now

catchingash committed:
Changed button display to indicate that this is the final submit order button

catchingash committed:
Persists invalid input to display on the form when the user receives an error.

Better UX if they don't have to re-enter valid input, and if they can see the invalid input to see what they did wrong.

catchingash committed:
Requires form fields, and changes types to match input.

sojeri committed:
added #remove_stock, #add_stock, and :has_stock to product model & specs

corinnepingul committed:
WIP.

lindseyreno committed:
Adds the ability for the product show page to see the categories

lindseyreno committed:
Adds categories to the product show page

catchingash committed:
Shortens credit cards before validation / before saving to the DB.

Rewrote the relevant test to be multiple tests. This adds clarity when looking into why a test is failing or passing.

sojeri committed:
added #adjust_if_product_stock_changed! and #remove_product_stock! to OrderItem model & specs

sojeri committed:
fixing broken specs

corinnepingul committed:
Finished specs for session when there is a valid user.

sojeri committed:
fixed broken specs

corinnepingul committed:
Finished specs for sessions controller post #create method.

corinnepingul committed:
added specs for DELETE #destroy.

sojeri committed:
added method to help cart know when to display + / increase qty links

corinnepingul committed:
Adding application_controller_spec.rb file.

sojeri committed:
fixed Order model's #prepare_checkout & updated orders controller & specs.

sojeri committed:
merged in tux/master

corinnepingul committed:
WIP.

sojeri committed:
why merge no fix conflict? ;_;

corinnepingul committed:
Changing Seller.new to Seller.create.

lindseyreno committed:
Lets the seller see the categories assigned to that product

sojeri committed:
added specs to test #prepare_checkout! and #checkout!(checkout_params)

lindseyreno committed:
Rewrites the categories routes so only a seller can create a category

lindseyreno committed:
Adds ability for a seller to create a category

lindseyreno committed:
Adds a link to create a new category

sojeri committed:
minor fixes per comments on PR

catchingash committed:
Merge pull request #48 from catchingash/tux/js/update-stock

Changes how avail stock is calculated; also adds a check to checkout.

corinnepingul committed:
WIP.

corinnepingul committed:
Changing second creation.

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/checkout_form_updates

catchingash committed:
Merge pull request #46 from catchingash/tux/crp/sessions_controller_specs

Sessions controller specs

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/checkout_form_updates

lindseyreno committed:
Merge pull request #47 from catchingash/tux/anw/checkout_form_updates

Checkout form updates

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/create_assign_categories

sojeri committed:
updated products #index & product #self.top_products to use :has_stock scope

sojeri committed:
added retired boolean column to products table

corinnepingul committed:
Added application controller specs.

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/application_controller_specs

sojeri committed:
added retire button to product show page

sojeri committed:
added retire button to seller products page

catchingash committed:
Adds `ruby '2.2.2'` to Gemfile so we can avoid the warning!

lindseyreno committed:
Adds ability to add categories to a product

catchingash committed:
Merge pull request #49 from catchingash/tux/crp/application_controller_specs

Application controller specs

sojeri committed:
added specs for #retire! & updated :has_stock & #top_products specs to account for updated code

sojeri committed:
pulling in tux/master in case it'd change anything

sojeri committed:
added specs for products helper #retire_product_text

sojeri committed:
updated product controller specs

sojeri committed:
added a few more specs for #edit, #update, & #retire

lindseyreno committed:
Adds tests for creating a new category.  All tests pass.

lindseyreno committed:
Removes unused flash message

sojeri committed:
added FIXME for bug I just noticed

sojeri committed:
added a couple more specs to fix outstanding FIXMEs I'd forgotten. I <3 rake notes.

lindseyreno committed:
Adds tests for adding categories to a product

lindseyreno committed:
Fixes #add_categories to be able to handle adding a single category to a product as well as multiple categories

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/create_assign_categories

lindseyreno committed:
Fixes test so extra code in controller not needed and deleted.  All tests pass.

lindseyreno committed:
Redirects a logged in seller away from the signup page

lindseyreno committed:
Adds passing test for redirecting a logged in seller away from the signup page

sojeri committed:
added more product seeds

sojeri committed:
added some more review seeds but something wrong with them

catchingash committed:
Merge pull request #53 from catchingash/tux/js/retire-product

Retire product

catchingash committed:
Adds status column to order_items

catchingash committed:
When an order is marked as paid, the order_items will also be marked as paid.

Also changed `item` to `order_item` to match other codebase conventions.

catchingash committed:
Renames `items` to `order_items` to match codebase conventions.

sojeri committed:
seeds update is ready for PR :smile:

catchingash committed:
Adds functionality to mark an order_item as shipped or canceled.

catchingash committed:
Fixes some bugs that resulted from not having a valid OrderItem instance

catchingash committed:
Fixes bug where empty messages will be displayed.

catchingash committed:
Removes unnecessary comments.

catchingash committed:
Bug fix: uniqueness is checked ony on create.

Prior, an order_item would not properly update.

catchingash committed:
Adds validation for new OrderItem status column

catchingash committed:
Merge pull request #54 from catchingash/tux/js/seeds

Expands seed data

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/order_item_archive

catchingash committed:
Minor refactor: Simplified seeds.rb.

corinnepingul committed:
Made the thumbnails a bit larger than before.

corinnepingul committed:
Changed thumbnails so they display the full product description if it's less than 150 characters.

corinnepingul committed:
Changed buttons and seller dashboard...a little :)

lindseyreno committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/lr/restrict_access_to_signup

lindseyreno committed:
Refactors new action

catchingash committed:
Merge pull request #52 from catchingash/tux/lr/restrict_access_to_signup

Restrict access to signup page

corinnepingul committed:
Merge pull request #55 from catchingash/tux/anw/order_item_archive

Order item ship / cancel

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/more_styling

catchingash committed:
Merge pull request #56 from catchingash/tux/crp/more_styling

More styling

corinnepingul committed:
OPTIMIZE comments addressed/deleted in applicaiton controller file.

lindseyreno committed:
Changed routes so categories new and create are not nested and made the rest of the code work with the new routes.  All tests pass.

corinnepingul committed:
TODO message deleted in application controller. I wrote the tests for the two methods yesterday.

lindseyreno committed:
Changes password_digest to password and password_confimation

corinnepingul committed:
TODO and OPTIMIZE comments addressed. Will write test for the last TODO remaining.

catchingash committed:
Fixes issue where emtpy error arrays will be displayed.

catchingash committed:
Adds credit card CVV form field.

Also adds some placeholder text to the other fields.

lindseyreno committed:
Fixes merge conflicts

catchingash committed:
Merge pull request #51 from catchingash/tux/lr/create_assign_categories

Create assign categories

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/checkout_form_updates

lindseyreno committed:
Merge pull request #57 from catchingash/tux/anw/checkout_form_updates

Checkout form updates

corinnepingul committed:
added test for updating an order item's status to paid after checkout.

lindseyreno committed:
Updates receipt to include project requirements

catchingash committed:
Removes unnecessary comments.

corinnepingul committed:
Added spec tests for the rest of order_item model.

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/addressing_comments

lindseyreno committed:
Deletes unused helper and helper_spec files

corinnepingul committed:
I left this one optimization in because I think someone else should look at it and decide if we want to move this to another controller.

corinnepingul committed:
Merge pull request #59 from catchingash/tux/lr/receipt

Updates receipt to include project requirements

corinnepingul committed:
Merge pull request #60 from catchingash/tux/lr/delete_unused_files

Deletes unused helper and helper_spec files

lindseyreno committed:
Merge pull request #61 from catchingash/tux/crp/addressing_comments

Tux/crp/addressing comments

corinnepingul committed:
Changed categories buttons to not have a border.

corinnepingul committed:
Changed buttons retire/reactivate to say only one word..they were too long before

sojeri committed:
added wrapper to require_seller_login for products#show to handle not displaying retired products to unauthenticated users

sojeri committed:
merging in tux/master. fixing product access now ready for PR.

corinnepingul committed:
Moved add a category button to the seller's dashboard.

corinnepingul committed:
Made sure each page header is styled the same. Edited some buttons.

sojeri committed:
updated sellers #show to use :has_stock scope

sojeri committed:
updated categories #show to use :has_stock scope

catchingash committed:
Updates seller pages per requirements.

On the order fulfillment page:
  - Total Revenue
  - Total Revenue by status
  - Total number of orders by status
  - Filter orders displayed by status
  - Link to each individual order
  - A list of orders including at least one of my products: A link to the item description page
  - DateTime the order was placed

corinnepingul committed:
Changed product show page a bit.

catchingash committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/anw/seller_orders_page_updates

lindseyreno committed:
Merge pull request #63 from catchingash/tux/anw/seller_orders_page_updates

Seller orders page updates

corinnepingul committed:
Won't show categories if there aren't any associated with that product.

corinnepingul committed:
Put categories and reviews on different rows.

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/even_more_styling

catchingash committed:
Merge pull request #64 from catchingash/tux/crp/even_more_styling

Added more styling to make site more consistent

sojeri committed:
updated seller specs

sojeri committed:
merging in tux/master

catchingash committed:
Makes changes so that out-of-stock but still active products are displayed.

lindseyreno committed:
Merge pull request #62 from catchingash/tux/js/fixing-retire-access

fixing retire access

corinnepingul committed:
Added page headers to views.

corinnepingul committed:
Merge branch 'tux/master' of github.com:catchingash/C3Projects--bEtsy into tux/crp/last_styling

lindseyreno committed:
Merge pull request #65 from catchingash/tux/crp/last_styling

Last styling things --> added page headers to views

catchingash committed:
Bug fix: "shipped" was not a valid status

sojeri committed:
added tux favicon

sojeri committed:
merging in tux/master

catchingash committed:
Merge pull request #66 from catchingash/tux/js/favicon

added tux favicon

kariabancroft commented on app/controllers/categories_controller.rb:
```diff
@@ -1,15 +1,35 @@
 class CategoriesController < ApplicationController
   before_action :set_category, only: [:show]
```
Was this created with the thought that it would be needed somewhere else? It seems a bit odd that it is only used for one method.

catchingash replied:
This was how we handled this in other controllers so we wanted to do it this way for consistency, and to make it easy later on to add other actions.

kariabancroft commented on app/controllers/order_items_controller.rb:
```diff
@@ -1,82 +1,40 @@
-    if @item.quantity_ordered == 1
-      flash[:error] = "You cannot decrease the quantity of #{ @item.display_name } any further. You must remove it from your cart."
+  def update
+    if params[:ship] == "true"
```
Why did you decide to use the string `"true"` rather than a boolean T/F?

catchingash replied:
We made this a string because the params are strings, so `true` doesn't work.

kariabancroft commented on app/controllers/order_items_controller.rb:
```diff
@@ -1,82 +1,40 @@
-    def flash_more_error # we should talk about whether or not this error message should be displayed & what it should truly say in production
-      flash[:error] = "You cannot increase the quantity of #{ @item.display_name } any further, because there are only #{ @item.quantity_ordered } in stock."
+    def set_order_item
+      @order_item = OrderItem.find_by(id: params[:id])
```
`find_by` will return `nil` if there is no corresponding object in the database. What will happen down the line in the methods that use this `@order_item` if it is set to `nil`?

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/controllers/orders_controller.rb:
```diff
@@ -1,48 +1,75 @@
-  before_action :find_order, only: [:cart, :checkout, :update, :receipt]
-  before_action :check_access, only: [:cart, :checkout, :receipt]
-
+  before_action :set_order, only: [:cart, :checkout, :add_to_cart, :update, :receipt]
```
It would be more effective to use the `except` on this rather than the `only`

catchingash replied:
We know that it's more efficient, but we wanted to avoid bugs down the road if someone added a new action to the controller and didn't see this.

kariabancroft commented on app/controllers/orders_controller.rb:
```diff
@@ -1,48 +1,75 @@
+  before_action :set_seller, only: [:index, :show]
+  before_action :require_seller_login, only: [:index, :show]
 
   def cart; end
```
Nice! This is the best way, according to our style guide, to use methods that have no body

catchingash replied:
:+1: :+1: :+1: :+1:  (From all four of us :smile: )

kariabancroft commented on app/controllers/orders_controller.rb:
```diff
@@ -1,48 +1,75 @@
-      order_info[:status] = "paid"
-
-      return order_info
+      params.permit(order: [:buyer_name, :buyer_email, :buyer_address, :buyer_card_short, :buyer_card_expiration])[:order]
```
Are any of these required?

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/controllers/products_controller.rb:
```diff
@@ -1,29 +1,76 @@
+      params.require(:product).permit(:name, :price, :stock, :description, :photo_url)
+    end
+
+    def require_seller_login_if_retired
```
This method seems unnecessary since it's only used in the show and it could be re-written in a single-line `if`statement

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,2 +1,28 @@
 class SessionsController < ApplicationController
+  before_action :set_seller, only: [:create]
+
+  def new
```
This should be using the single-line syntax that the other methods with no body are using

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/models/order.rb:
```diff
@@ -1,36 +1,88 @@
     end
+  end
+
+  def total_order_price(seller_id=nil)
```
Add spaces around the `=`

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/models/order.rb:
```diff
@@ -1,36 +1,88 @@
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  VALID_BUYER_CARD_SHORT_REGEX = /\A\d{4}\z/
+
+  validates :status, presence: true, inclusion: { in: %w(pending paid complete canceled),
```
I think it would make sense to create a constant or constants that correspond to these statuses because you use them later on

catchingash replied:
:+1: x4

kariabancroft commented on app/models/order_item.rb:
```diff
@@ -1,37 +1,70 @@
+  end
+
+  private
+    def order_item_is_unique?
```
It doesn't seem like this method is actually a `?` method because it doesn't seem to be returning a true or false

catchingash replied:
:+1:

kariabancroft commented on app/models/product.rb:
```diff
@@ -4,25 +4,59 @@ class Product < ActiveRecord::Base
-  def self.average_rating(id)
-    product = Product.find(id)
+  scope :active, -> { where(retired: false) }
+  scope :has_stock, -> { where(retired: false).where("stock > 0") }
```
Can you use the `active` scope above rather than repeating the logic?

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/helpers/application_helper.rb:
```diff
@@ -2,4 +2,23 @@ module ApplicationHelper
+    display_text = "Cart"
+
+    if session[:order_id] && Order.find_by(id: session[:order_id])
+      no_items = Order.find(session[:order_id]).order_items.count
```
I would reduce the amount of chaining on this line - too many object/type leaps

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/models/order_item.rb:
```diff
@@ -1,37 +1,70 @@
+    if quantity_ordered > 1
+      update_column(:quantity_ordered, quantity_ordered - 1)
+    else
+      errors.add(:quantity_ordered, "You must remove this from your cart if you want to reduce its quantity any further.")
```
There are a number of error message strings used in your models, these would be great to put into constants and/or all in one array to keep separate from the rest of your code.

catchingash replied:
:+1:

kariabancroft commented on app/models/order_item.rb:
```diff
@@ -1,37 +1,70 @@
     quantity_ordered * product.price
   end
+
+  def product_has_stock?
```
It seems like this is a method that should be on the product model not the OrderItem model

catchingash replied:
:+1: We had more in this method originally, and then removed the extra functionality... but by then, we'd called it in many places.

kariabancroft commented on app/models/product.rb:
```diff
@@ -4,25 +4,59 @@ class Product < ActiveRecord::Base
   end
+
+  def stock?
+    stock > 0
```
It seems like this method is serving the same purpose as the `has_stock` scope. Can the logic be consolidated?

catchingash replied:
The scope returns all products that have stock, whereas this one returns true if a specific product has stock.

kariabancroft commented on app/views/orders/index.html.erb:
```diff
@@ -0,0 +1,80 @@
+        <th>Total</th>
+      </tr>
+    </thead>
+     <tr>
```
I wonder if this whole section could be cleaned up with an array that contained the statuses since the rest of the markup is the same

catchingash replied:
:+1: x4

We wonder things sometimes too. (Kidding!!! Have a great weekend :smile_cat: )

kariabancroft commented on app/views/products/edit.html.erb:
```diff
@@ -0,0 +1,18 @@
+    <h3>Assign Categories</h3>
+    <%= form_for @product, url: product_add_categories_path(@product.id), html: { class: "form" } do |f| %>
+      <% @categories.each do |category| %>
+      <div class="checkboxes">
```
Should this div be inside or outside the `each`?

catchingash replied:
:heavy_check_mark:

kariabancroft commented on app/views/reviews/new.html.erb:
```diff
@@ -1,14 +1,25 @@
+</div>
+<div class="row">
+  <div class="col-lg-4 col-lg-offset-4">
+    <%= form_for @review, url: "/products/#{@product.id}/reviews/new" do |f| %>
```
Is there a named route to use in place for this?

catchingash replied:
:heavy_check_mark:

kariabancroft commented on spec/controllers/application_controller_spec.rb:
```diff
@@ -0,0 +1,76 @@
+
+  describe "require_seller_login" do
+    before :each do
+      @seller1 = Seller.create(username: "user1", email: "email1@email.com", password: "password1", password_confirmation: "password1")
```
It seems like this `@seller1` variable is not needed in the overall before each, but rather should be in the before each in the "seller logged in" section

catchingash replied:
:heavy_check_mark:

kariabancroft commented on spec/controllers/application_controller_spec.rb:
```diff
@@ -0,0 +1,76 @@
+  describe "set_seller" do
+    before :each do
+      @seller1 = Seller.create(username: "user1", email: "email1@email.com", password: "password1", password_confirmation: "password1")
+      @seller2 = Seller.create(username: "user2", email: "email2@email.com", password: "password2", password_confirmation: "password2")
```
I don't think the `@seller2` is needed in before each since it appears as though it is only used in one of the tests below

catchingash replied:
:heavy_check_mark:

kariabancroft commented on spec/controllers/categories_controller_spec.rb:
```diff
@@ -48,7 +52,8 @@
         create_11_categories
```
This test and tests below appear to be describing what your code is doing rather than what the expected behavior is. Try to avoid describing exactly what the code is doing and focus on the behaviors.

catchingash replied:
:+1: x4

kariabancroft commented on spec/controllers/categories_controller_spec.rb:
```diff
@@ -73,9 +78,90 @@
+          seller
+          session[:seller_id] = seller.id
+
+          post :create, valid_params
```
The first three lines of code in this test are exactly the same as the first three lines in all of the tests below. You should create a separate method that does this same operation each time rather than duplicating the test code every time.

catchingash replied:
:heavy_check_mark:

kariabancroft commented on spec/controllers/orders_controller_spec.rb:
```diff
@@ -69,97 +53,368 @@
+      session[:order_id] = test_order.id
+    end
+
+    it "assigns @order" do
```
Here is another set of tests where it would be preferred to describe the behaviors rather than what the specific code is doing

catchingash replied:
:+1: x4

kariabancroft commented on spec/controllers/orders_controller_spec.rb:
```diff
@@ -69,97 +53,368 @@
+      let(:test_order_canceled) { Order.create( status: "canceled", buyer_name: "My name", buyer_email: "my_email@example.com", buyer_address: "123 Example St, Cityville, State 12345", buyer_card_short: "1234", buyer_card_expiration: future_date ) }
+
+
+      it "redirects to root when 'pending'" do
```
It seems like these three tests would be nicely combined by putting them each in an array and then iterating over that and executing the same test for each item in the array.

catchingash replied:
:+1: x4

kariabancroft commented:
There are a few additional things that you can clean up overall. For products, categories and sellers that use the IDs in the URL, you will want to handle whether or not the ID that was supplied exists, for example if a user hits the URL `/sellers/99876`, what will happen?

kariabancroft commented:
@catchingash, @drvonnjerryxlii, @lindseyreno, and @corinnepingul --
Very nice job on this! Your tests are extremely thorough. I think with your newly discovered knowledge of FactoryGirl you could really clean up your tests even more. There are certainly a few things in your code base that show me that this was developed by four different people, which is to be expected. There was one specific example that I noticed where it seems like duplicate functionality was created in two different ways (presumably by two different people). 
I like the way that you displayed the `-` and `+` on the quantity items in the cart. Initially I was confused about why I can two items in my cart and couldn't add another. To enhance this user experience, you could still display the `+`, not as a link, with a hover-over text stating that there are no more items in stock. Future fun!
Great job!

catchingash committed:
Changes an unnecessary find_by to find.

Specs still green :)

`find` will raise an error if the order_item cannot be found.

catchingash committed:
Minor refactor.

Doesn't change any functionality!

catchingash committed:
Removes unnecessary method.

catchingash committed:
Minor refactor

No functionality changes :)

catchingash committed:
Minor refactor: reduces number of chained methods

catchingash committed:
Fixes typo!

*tucks tail*

catchingash committed:
Adds spaces around `=`

catchingash committed:
Removes misleading `?` from method name

catchingash committed:
Minor refactor

No functionality changes

catchingash committed:
Updates "checkboxes" --> "checkbox"

Makes this more clear, since this is surrounding a single checkbox.

catchingash committed:
Updates "checkboxes" --> "checkbox"

Makes this more clear, since this is surrounding a single checkbox.

catchingash committed:
Changes to a dynamic path instead of static

catchingash committed:
Refactor: sellers were being created when not used

catchingash committed:
DRY: Reduces duplication :)

catchingash commented:
Awesome comments! Thanks for your help, @kariabancroft !

sojeri committed:
added a few specs

sojeri committed:
last two specs done. i hope you liek 100% coverage, next group. ;)
